### DISCUSSION TITLE
How to access the current i18n locale in an API route?

### DISCUSSION THREAD

**Discussion by paulobarcelos** [vercel/next.js#21798] [Help]
When using i18n routes (https://nextjs.org/docs/advanced-features/i18n-routing), I can easily figure out what is the current locale in `getStaticProps({ locale })`, however, when inside and API route, I simply can't/

I used to be able to use `req.query.__nextLocale`, but is not available anymore after 10.0.5.

Tried to check cookies, try to import the router instance from `next/router` and nothing works.
Even tried to see if I can parse the url and figure the locale from it, but if I for example access `http://localhost:3000/en-US/api/test`, then `req.url` is `/api/test` and not `/en-US/api/test`... so the locale is hidden even there.

This is very frustrating...
If the locale is available on getStaticProps I don't understand why that information shouldn't be exposed on an API route...

**Comment by smessaris** [comment]
Having the same problem, is there a solution?

**Reply by paulobarcelos** [reply]
I couldn't find one. Ended up having to pass the locale via a request
header.

**Comment by Linux249** [comment]
Just stumbled across the problem too - this somehow look's like a half implementation. First I start to implement to have a look if it works -> it works ("redirecting" /en-US/api/... -> to /api/...) but then I didn't found a way to get access to the "en-US". 

Question: Why is next.js removing the information from the `url` but don't give access to the information?

**Reply by Eunit99** [reply]
I have exactly this same question.

**Comment by aralroca** [comment]
It's not the best solution, but If you save the `NEXT_LOCALE` on `document.cookie` you can use `req.cookies.NEXT_LOCALE`

https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie

**Comment by Raggaer** [comment]
Any news about this? I end up sending the locale on the API request.

**Comment by orneryd** [comment]
here is what I ended up doing to get the locale. I run a custom express server with  a use handler:


```javascript 
server.use((req, res, next) => {
        req.locale = req.query.locale || detectLocale(req.originalUrl) || ("en");
        req.client?.setAcceptLanguage(req.locale)
        res.set("Content-Language", req.locale)
        next()
    });
```
I import the list of locales also imported into next.config.js for the detectlocale function. This function is also much more efficient than the next.js detect locales function.

```javascript
import {locales} from "../i18n";
const localeList = locales.reduce((acc, locale)=> {
    acc[locale] = locale.toLowerCase();
    return acc;
}, {})

/**
 * For a pathname that may include a locale from a list of locales, it
 * removes the locale from the pathname returning it alongside with the
 * detected locale.
 *
 * @param pathname A pathname that may include a locale.
 * @returns The detected locale and pathname without locale
 */
export function detectLocale(pathname: string) {
    const pathnameParts = pathname.split('/');
    // first item will be empty string from splitting at first char
    return localeList[(pathnameParts[1] || "").toLowerCase()];
}

**Comment by Dragate** [comment]
My current workaround taking into account NEXT_LOCALE.

lib/common.ts:
```ts
import type { NextApiRequest } from "next"
import { acceptLanguage } from "next/dist/server/accept-header"
import { i18n } from "next.config"

export function currentLocale(req: NextApiRequest): string {
    if (!i18n) throw new Error("i18n not configured")
    const chosenLocale = i18n.locales.find(locale => locale == req.cookies.NEXT_LOCALE)
    const detectedLocale = chosenLocale ?? acceptLanguage(req.headers["accept-language"], i18n.locales)
    return detectedLocale || i18n.defaultLocale
}
```
pages/api/test.ts:
```ts
import { currentLocale } from "lib/common"
import type { NextApiRequest, NextApiResponse } from "next"

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
    const { method } = req

    const locale = currentLocale(req)

    switch (method) {
        case "GET": {
            return res.json(locale)
        }
        default:
            return res.status(405).end()
    }
}
```

**Reply by jeffcad** [reply]
I just pass it in the request body object, and didn't think anything of it. It's pretty easy, so is there a reason that others are using more complex solutions?

**Reply by Dragate** [reply]
> I just pass it in the request body object, and didn't think anything of it. It's pretty easy, so is there a reason that others are using more complex solutions?

Preference mostly. I don't want to increase my client-side code when the locale information already exists in the headers of each request.

**Comment by radikris** [comment]
Can anyone please check out my question here, it is related to the question, and I am totally stucked with it.
https://stackoverflow.com/questions/72638109/nextjs-api-folder-call-500-error-in-production

**Comment by nbouvrette** [comment]
I'm not sure APIs should have locales given that they are under the `API` path (no locales defined in the path). Instead, you can simply specify the locale when you call the API like this:

```ts
fetch(`/api/hello`, {
  headers: { 'Accept-Language': locale }
})...
```

See the full example here: https://github.com/Avansai/next-multilingual/blob/5b2875192ea06fee6969badc6cbee0c260c2b699/example/pages/index.tsx#L52

**Comment by carlosrsabreu** [comment]
Any updates regarding this? Is there any official solution?

**Reply by YHADJRABIA** [reply]
Hey man, I put an update below if needed

**Comment by YHADJRABIA** [comment]
Most recent solution using **Next 14**'s `app` router with `next-intl` package:

The active locale is automatically stored in a `NEXT_LOCALE` cookie, which can be accessed from the api route as shown below, no need to resort to request headers.

![image](https://github.com/vercel/next.js/assets/39564199/8c85df8e-b7fb-46a4-b97a-2bc1b433c932)


```
export async function POST(req: NextRequest) {
  try {
    const nextLocaleCookie = req.cookies.get('NEXT_LOCALE') // Reading the cookie: { name: 'NEXT_LOCALE', value: 'en'}
    const locale = nextLocaleCookie ? nextLocaleCookie.value : yourDefaultLocale // Extracting locale's value: 'en'
   ...
```

**Comment by mrmgreen** [comment]
You can use a rewrite to pass the locale as a query param. 
Which you can access in the api using `req.query`.

Example using sub path locale routing. 
`  {
    source: '/:locale/robots.txt',
    destination: '/api/robots?locale=:locale',
    locale: false,
  },`

**Reply by brunel-pierrebakker** [reply]
This is working for me. thnx!
